test/Test/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,525 @@
+    lastKeysBlock = Primitive.concat (lastKeysBlocks lastKeys)
+
+{-
+    Incrementally constructs an index, using the functions 'new', 'append', and
+    'unsafeEnd'.
+-}
+incrementalConstruction :: [Append] -> (IndexOrdinary, Primitive.Vector Word8)
+incrementalConstruction appends = runST $ do
+    acc <- new keyCount minChunkSize
+    commonChunks <- mapM (flip append acc) appends
+    (remnant, unserialised) <- unsafeEnd acc
+    let
+
+        serialised :: Primitive.Vector Word8
+        serialised = Primitive.concat $
+                     fromChunk <$> catMaybes (commonChunks ++ [remnant])
+
+    return (unserialised, serialised)
+    where
+
+    keyCount :: Int
+    keyCount = sum (map appendedKeysCount appends)
+
+    {-
+        We do not need to vary the minimum chunk size, since we are not testing
+        chunk building here.
+    -}
+    minChunkSize :: Int
+    minChunkSize = 0x1000
+
+-- * Properties to test
+
+-- ** Search
+
+prop_searchForMentionedKeyWorks :: NonNegative (Small Int)
+                                -> SearchableIndex
+                                -> Property
+prop_searchForMentionedKeyWorks (NonNegative (Small pageNo))
+                                (SearchableIndex index)
+    = pageNo < length lastKeys ==> all (<  key) prefix    .&&.
+                                   all (== key) selection .&&.
+                                   all (>  key) suffix
+    where
+
+    lastKeys :: Vector SerialisedKey
+    lastKeys = fromIndexOrdinary index
+
+    key :: SerialisedKey
+    key = lastKeys ! pageNo
+
+    prefix, selection, suffix :: Vector SerialisedKey
+    (prefix, selection, suffix) = searchPartitioning key index
+
+prop_searchForUnmentionedKeyInRangeWorks :: SerialisedKey
+                                         -> SearchableIndex
+                                         -> Property
+prop_searchForUnmentionedKeyInRangeWorks key (SearchableIndex index)
+    = key `notElem` lastKeys && key <= last lastKeys ==>
+      key < selectionHead                     .&&.
+      all (<  key)           prefix           .&&.
+      all (== selectionHead) (tail selection) .&&.
+      all (>  selectionHead) suffix
+    where
+
+    lastKeys :: Vector SerialisedKey
+    lastKeys = fromIndexOrdinary index
+
+    prefix, selection, suffix :: Vector SerialisedKey
+    (prefix, selection, suffix) = searchPartitioning key index
+
+    selectionHead :: SerialisedKey
+    selectionHead = head selection
+
+prop_searchForUnmentionedKeyBeyondRangeWorks :: SerialisedKey
+                                             -> SearchableIndex
+                                             -> Property
+prop_searchForUnmentionedKeyBeyondRangeWorks key (SearchableIndex index)
+    = not (null lesserLastKeys) ==> all (<  selectionHead) prefix           .&&.
+                                    all (== selectionHead) (tail selection) .&&.
+                                    all (>  selectionHead) suffix
```

Comment:
Ah, at first I didn't see how that ensure that the key is unmentioned and beyond range, but we construct new index to have that. Makes sense.

But why do we still check the same properties? Couldn't we just say that `prefix === lesserLastKeys` and the rest is empty?

Reply:
No, because what should be returned is some arbitrary but valid page span, and so we have to check that a valid page span is indeed returned. Actually, the selection being empty would mean that the page span is not valid. By “valid” I mean here that the page span comprises a non-overflow page plus any overflow pages that belong to it.

Reply:
> Ah, at first I didn't see how that ensure that the key is unmentioned and beyond range, but we construct new index to have that. Makes sense.

This is my little trick to avoid the problem that the initial solution had: it used a conditional property to only consider those cases where the key was large enough by accident, and this resulted in too many test cases being discarded.

Reply:
> Actually, the selection being empty would mean that the page span is not valid

Ah, you're right. There is of course still a selection. We could still have more specific (and tighter?) conditions we check here, but there is an argument to be made for using the same as in the other cases.

Reply:
I don’t think that we can make the condition tighter. Currently, it says that the page span must be valid, and this is really all we can require.